zerver/tests/test_message_fetch.py

```diff
@@ -593,31 +592,243 @@ def _build_query(self, term: Dict[str, Any]) -> Select:
+                flags=[],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={"type": "private"},
+                flags=[],
+            )
+        )
+        self.assertFalse(
+            narrow_filter(
+                message={"display_recipient": "devel", "type": "stream", "subject": "java"},
+                flags=[],
+            )
+        )
+        self.assertFalse(
+            narrow_filter(
+                message={"display_recipient": "social", "type": "stream"},
+                flags=[],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["sender", "hamlet@zulip.com"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={"sender_email": "hamlet@zulip.com"},
+                flags=[],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={"sender_email": "cordelia@zulip.com"},
+                flags=[],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "dm"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={"type": "private"},
+                flags=[],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={"type": "stream"},
+                flags=[],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "private"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={"type": "private"},
+                flags=[],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={"type": "stream"},
+                flags=[],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "starred"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={},
+                flags=["starred"],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={},
+                flags=["alerted"],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "alerted"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={},
+                flags=["mentioned"],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={},
+                flags=["starred"],
+            )
+        )
+
+        ###
+
+        narrow_filter = build_narrow_filter([["is", "mentioned"]])
+
+        self.assertTrue(
+            narrow_filter(
+                message={},
+                flags=["mentioned"],
+            )
+        )
+
+        self.assertFalse(
+            narrow_filter(
+                message={},
+                flags=["starred"],
+            )
+        )
```

Comment:
This could maybe benefit from a `wildcard_mentioned` flag case if we wanted to improve the coverage.